### PR TITLE
[SHELL32] DefView must set CMF_EXTENDEDVERBS, CMF_CANRENAME and CMF_EXPLORE

### PR DESCRIPTION
**Pull request by whindsaks**
The caller of IContextMenu::QueryContextMenu must set these flags when needed, not CDefaultContextMenu.

Notes:
 - CMF_CANRENAME is always set by DefView because it always supports rename in its current form. CDefaultContextMenu verifies that SFGAO_CANRENAME is also set on the items. All other callers are expected to not know how to rename unless they also set CMF_CANRENAME. This fixes the bug that the Rename item is present in the TreeView context menu even though the TreeView in ROS Explorer does not handle renaming.
 - While DefView now tries to set CMF_EXPLORE correctly, the flag is never actually set because BROWSEUI does not handle FCW_TREE yet.
 - This PR also fixes the bug where the File menu is missing the menu items when there is no selection.

### CODE REVIEW

File: dll/win32/shell32/CDefView.cpp [hunk 1]
```diff
@@ -1511,10 +1527,10 @@ HRESULT CDefView::InvokeContextMenuCommand(CComPtr<IContextMenu>& pCM, LPCSTR lp
     cmi.hwnd = m_hWnd;
     cmi.lpVerb = lpVerb;
 
-    if (GetKeyState(VK_SHIFT) & 0x8000)
```
Comment (TAN-Gaming) [inline dll/win32/shell32/CDefView.cpp:1514]: What's wrong with `0x8000`?

Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:1514]: Comparing a value to 0 is better than testing if a bit is set.

Comment (TAN-Gaming) [reply dll/win32/shell32/CDefView.cpp:1514]: However, MS docs determine the returned value by "which bit is set".
[https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getkeystate](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getkeystate)

The example also uses `0x8000` to check if the key is pressed.
[https://learn.microsoft.com/en-us/windows/win32/inputdev/using-keyboard-input](https://learn.microsoft.com/en-us/windows/win32/inputdev/using-keyboard-input)

File: dll/win32/shell32/CDefView.cpp [hunk 2]
```diff
@@ -54,6 +54,20 @@ typedef struct
...
 
+static UINT
+GetContextMenuFlags(IShellBrowser *pSB, SFGAOF sfgao)
+{
+    UINT cmf = CMF_NORMAL;
+    if (GetKeyState(VK_SHIFT) < 0)
+        cmf |= CMF_EXTENDEDVERBS;
+    if (sfgao & SFGAO_CANRENAME)
+        cmf |= CMF_CANRENAME;
+    HWND hwnd = 0;
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:65]:
```suggestion
    HWND hwnd = NULL;
```

File: dll/win32/shell32/CDefView.cpp (hunk 1 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:1514]: > MS docs determine the status by checking "which bit is set"

To be pedantic; the docs describe the 4 possible return values in terms of how the bits are set, not how the caller should check them.

Surely it is no coincidence that the function returns a signed number and the state you are most interested in just happens to be the sign bit?

Comment (HBelusca) [reply dll/win32/shell32/CDefView.cpp:1514]: https://godbolt.org/z/f5G3bW7EW

Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:1514]: > https://godbolt.org/z/f5G3bW7EW

You should not let the compiler see the functions you are calling there.

Anyway, GCC is smart enough to optimize checking bit 0x8000 but MSVC is not:

https://godbolt.org/z/f3v67dddT

```
call    short GetKeyState(int)                    ; GetKeyState
cwde
test    eax, eax
jge ...
```

vs 

```
call    short GetKeyState(int)                    ; GetKeyState
cwde
and     eax, 32768                                ; 00008000H
test    eax, eax
je ...
```

Comment (TAN-Gaming) [reply dll/win32/shell32/CDefView.cpp:1514]: > Surely it is no coincidence that the function returns a signed number and the state you are most interested in just happens to be the sign bit?

I think following the docs is better than relying on assumptions.

Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:1514]: > I think following the docs is better than relying on assumptions.

Which part of the GetKeyState documentation is the code not following? It checks the bit we are interested in.

[Here the OldNewThing checks with `< 0`](https://devblogs.microsoft.com/oldnewthing/20040924-00/?p=37753) for the exact CMIC flags we are talking about...

Comment (TAN-Gaming) [reply dll/win32/shell32/CDefView.cpp:1514]: If it's really better for MSVC and doesn't break anything, fine then.

**Review by learn-more — APPROVED** [review]

File: dll/win32/shell32/CDefView.cpp (hunk 2 above)
Comment (HBelusca) [reply dll/win32/shell32/CDefView.cpp]: (re: "style says NULL and not zero": The underlying reason is because a handle is like a pointer.)

**Review by TAN-Gaming — APPROVED** [review]
Approved. Works fine when combined with my PR [#6725](https://github.com/reactos/reactos/pull/6725) .
However, my justification is only based on the results, not the correctness of the code.

**Review by learn-more — APPROVED** [review]

**Pull request merged**